proof.go

```diff
@@ -465,6 +465,203 @@ func (proof Proof) VerifyInclusion(h hash.Hash, nid namespace.ID, leavesWithoutN
 	return res
 }
 
+// VerifySubtreeRootInclusion verifies that a set of subtree roots is included in
+// an NMT.
+// Note: This method is Celestia specific.
+// It makes the following assumptions:
+// - The subtree roots are created according to the ADR-013
+// https://github.com/celestiaorg/celestia-app/blob/main/docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md
+// - The tree's number of leaves is a power of two
```

Comment:
this assumption is only for that method. The NMT implementation is general purpose

Reply:
It is not so obvious what happens if the tree doesn't follow that assumption and `VerifySubtreeRootInclusion` is called.

Reply:
[a68a69f](https://github.com/celestiaorg/nmt/pull/260/commits/a68a69ffb1108e22d8c1c2b9bb17c994806c7d0c)

Reply:
> // Warning: This method is Celestia specific! Using it without verifying
// the following assumptions, can return unexpected errors, false positive/negatives:

I think erroring is fine but false positives / negatives should not happen. Can't we assert that the passed-in params adhere to the assumptions and return an error if not?

Reply:
I'll look into it. I didn't add a check because it's not trivial to know the size of the tree from the proof. I'll rethink again if there is a way to do so.

Reply:
I am currently writing a fuzzer that tests a good number of combinations, and no true negatives are being reported. I will try to write a test that goes over all possible combinations and see.

Reply:
@liamsi The following test goes over a wide range of possible ranges using all the possible tree size from 1 to 255. Also, it goes over all the possible ranges of blobs and subtree root proofs and verifies them:
```go
func TestAllCombinations(t *testing.T) {
	type testCase struct {
		proof        Proof
		width        int
		tree         *NamespacedMerkleTree
		root         []byte
		subtreeRoots [][]byte
		shouldPass   bool
	}

	var testCases []testCase
	var nids = []byte{0}

	for numberOfNIDs := uint8(1); numberOfNIDs < 255; numberOfNIDs++ {
		nids = append(nids, numberOfNIDs)
		tree := exampleNMT(1, true, nids...)

		for i := uint8(0); i < numberOfNIDs; i++ {
			for j := i + 1; j < numberOfNIDs; j++ {
				shouldPass := true
				idx := NextShareIndex(int(i), int(j-i), 64)
				if idx != int(i) {
					// we will populate using random subtree roots in this case
					shouldPass = false
				}

				width := SubTreeWidth(int(j-i), 64)

				ranges, err := ToLeafRanges(int(i), int(j), width)
				require.NoError(t, err)
				proof, err := tree.ProveRange(int(i), int(j))
				require.NoError(t, err)

				var subtreeRoots [][]byte
				for _, rg := range ranges {
					s, err := tree.ComputeSubtreeRoot(rg.Start, rg.End)
					if err != nil {
						if !shouldPass {
							s = bytes.Repeat([]byte{1}, 32)
						} else {
							require.NoError(t, err)
						}
					}
					subtreeRoots = append(subtreeRoots, s)
				}

				root, err := tree.Root()
				require.NoError(t, err)

				testCases = append(testCases, testCase{
					proof:        proof,
					width:        width,
					tree:         tree,
					root:         root,
					subtreeRoots: subtreeRoots,
					shouldPass:   shouldPass,
				})
			}
		}

	}

	nmthasher := testCases[0].tree.treeHasher
	hasher := nmthasher.(*NmtHasher)

	for _, hi := range testCases {
		fmt.Printf("size:%d,start:%d,end:%d,width:%d\n", hi.tree.Size(), hi.proof.start, hi.proof.end, hi.width)
		result, err := hi.proof.VerifySubtreeRootInclusion(hasher, hi.subtreeRoots, hi.width, hi.root)
		if hi.shouldPass {
			assert.NoError(t, err)
			assert.True(t, result)
		} else {
			assert.Error(t, err)
			assert.False(t, result)
		}
		fmt.Println(result)
	}
}
```

I didn't include it in the code because it takes 4 min or so to execute since it's basically brute forcing all possibilities. Also, it uses code from go-square so it's not good to have these imported in nmt.

Now I guess the tree size condition is not necessary, only the width needs to be valid. And even if the width is not valid, it returns an error. So no false positives are returned up to the tree size of 255. 

Do you think this is enough to remove the warning and just say it will return an error if the conditions are not met? Since I brute forced all possibilities for all ranges from a tree of size 1 to 255?